lib/views/post/widgets/vote_button.dart

```diff
@@ -0,0 +1,170 @@
+import 'package:pr0gramm/services/vote_animation_service.dart';
+
+class VoteButton extends StatefulWidget {
+  final VoteAnimationService animationService;
+  final VoteButtonType type;
+  final bool disabled;
+
+  VoteButton({
+    Key key,
+    @required this.type,
+    this.disabled = false,
+    this.animationService,
+  }) : super(key: key);
+
+  @override
+  _VoteButtonState createState() {
+    return _VoteButtonState();
+  }
+}
+
+class _VoteButtonState extends State<VoteButton> with TickerProviderStateMixin {
+  AnimationController rotationController;
+  AnimationController colorController;
+  AnimationController favoriteController;
+  Animation<Color> colorTween;
+  Animation<double> favoriteMorph;
+  CurvedAnimation favoriteMorphCurveAnimation;
+  Color color;
+  Color lastColor;
+  IconData icon;
+
+  onStateChange(VoteAnimation voteAnimation) {
+    setState(() {
+      lastColor = color;
+      color = _getColorByAnimation(voteAnimation);
+
+      if (voteAnimation == VoteAnimation.voteFocused ||
+          voteAnimation == VoteAnimation.voteUnfocused) {
+        if (widget.type == VoteButtonType.favorite) {
+          if (widget.animationService.initialVote == Vote.favorite)
+            favoriteController.reverse();
+          else
+            favoriteController.forward();
+        } else
+          rotationController.forward();
+      } else if (voteAnimation == VoteAnimation.clearFocused ||
+          voteAnimation == VoteAnimation.clearUnfocused) {
+        if (widget.type == VoteButtonType.favorite) {
+          if (widget.animationService.initialVote != Vote.favorite)
+            favoriteController.reverse();
+          else
+            favoriteController.forward();
+        } else
+          rotationController.reverse();
+      }
+
+      if (lastColor != color) {
+        colorTween =
+            ColorTween(begin: lastColor, end: color).animate(colorController);
+        colorController.forward(from: 0);
+      }
+    });
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    lastColor = color = _getInitialColor();
+    rotationController = new AnimationController(
+      vsync: this,
+      duration: voteAnimationDuration,
+    );
+    colorController = new AnimationController(
+      vsync: this,
+      duration: voteAnimationDuration,
+    );
+    favoriteController = new AnimationController(
+      vsync: this,
+      duration: voteAnimationDuration,
+    );
+    widget.animationService.addButtonStateListener(widget.type, onStateChange);
+    icon = widget.type == VoteButtonType.up
+        ? Icons.add_circle_outline
+        : Icons.remove_circle_outline;
+    colorTween =
+        ColorTween(begin: lastColor, end: color).animate(colorController);
+    if (widget.type == VoteButtonType.favorite) {
+      icon = null;
+      favoriteMorphCurveAnimation = CurvedAnimation(
+          parent: favoriteController,
+          curve: Curves.easeOutCubic,
+          reverseCurve: Curves.easeInCubic);
```

Comment:
Add a trailing comma to improve formatting

Reply:
will be obsolete when splitting into multiple Widgets